File: Half_Subtractor/test_gen.c
// Test vector file generation using C language you can also use python to generate "TRACEFILE.txt
#include <stdio.h> 
#include <stdlib.h>
#include <string.h>
#include <math.h>

// file pointer as global
FILE *optr;

// user defined function to convert decimal to binary and print in TRACEFILE.txt
void dec_to_bin(long long num, int len);

// user defined function to convert binary to decimal for inputs
int bin_to_dec(long long num);

// function definition for decimal to binary conversion
void dec_to_bin(long long num, int len) {
    // array to store binary values (initial length set to 20, change it according to your design)
    int bin[20];
    int rem, i, j;
    i = 0;

    // simple binary conversion, pen paper logic
    while(num != 0) {
        rem = num % 2;
        num /= 2;
        bin[i] = rem;
        ++i;
    }

    // print binary values with MSB first (padded with extra 0s)
    for(j = len - 1; j >= 0; j--) {
        if(j < i)
            fprintf(optr, "%d", bin[j]);
        else
            fprintf(optr, "0");
    }
}

// function definition for binary to decimal conversion
int bin_to_dec(long long num) {
    int i, dec, rem;
    dec = 0;
    i = 0;
    // simple digit extraction logic
    while(num != 0) {
        rem = num % 10;
        num /= 10;
        dec += rem * pow(2, i);
        ++i;
    }
    return dec;
}

int main() {
    char output_file[20];
    int input_len, output_len, test_len;
    int i, a, b, output, input;
    
    // name the tracefile
    strcpy(output_file, "TRACEFILE.txt");
    // open TRACEFILE.txt in write mode
    optr = fopen(output_file, "w");

    // input and output vector length for HALF-SUBTRACTOR
    input_len = 2;
    output_len = 2;
    // total number of test vectors (exhaustive) = 2^(number of input bits)
    test_len = pow(2, input_len);

    // run loop for all possible combinations of input
    for(i = 0; i < test_len; i++) {
        input = i;
        // separate the inputs (binary form)
        a = (i >> 1) & 1;
        b = i & 1;

        // perform the operation (in decimal)
        int D= a ^ b;
        int BO = ((!a) & b);
        output = (BO << 1) | D;

        // print input in "TRACEFILE.txt" (in binary)
        dec_to_bin(input, input_len);

        // single space
        fprintf(optr, " ");

        // print output in "TRACEFILE.txt" (in binary)
        dec_to_bin(output, output_len);

        // single space
        fprintf(optr, " ");

        // mask bits (equal to number of output bits)
        fprintf(optr, "11");

        // change line
        fprintf(optr, "\n");
    }

    // close file (TRACEFILE.txt)
    fclose(optr);
    return 0;
}

File: Half_Subtractor/test_gen.py
# Python code for TRACEFILE.txt generation for HALF-SUBTRACTOR
# Open/create TRACEFILE.txt in write mode
f = open("TRACEFILE.txt", "w")

# Number of inputs
input_len = 2

# Number of outputs
output_len = 2

# Number of test vectors
test_len = 2 ** input_len

# Loop from 0 to (test_len-1)
for input_vec in range(test_len):
    # Convert input to binary (2-bit) format
    input_str = "{:02b}".format(int(input_vec))

    # Extracting the input bits
    a = (input_vec >> 1) & 1
    b = input_vec & 1

    # Generating the output
    D= a ^ b
    BO = ((~a) & b)
    output = (BO << 1) | D

    # Convert output to binary (2-bit) format
    output_str = "{:02b}".format(int(output))

    # Write input and output test vector with 2-bit mask in TRACEFILE.txt
    f.write(input_str + " " + output_str + " 11\n")

# Close file
f.close()

# Alternate python code to produce the same result choose according to your preference

# Function to convert decimal to binary with fixed length
# def dec_to_bin(num, length):
#     return format(num, f'0{length}b')

# # Main function to generate the trace file
# def generate_tracefile():
#     output_file = "TRACEFILE_HalfSubtractor.txt"
#     input_len = 2  # Input vector length for HALF-SUBTRACTOR (2 inputs: a, b)
#     output_len = 2 # Output vector length for HALF-SUBTRACTOR (2 outputs: diff, borrow)
#     test_len = 2 ** input_len  # Total number of test vectors (exhaustive)

#     with open(output_file, "w") as optr:
#         for i in range(test_len):
#             input_bin = dec_to_bin(i, input_len)  # Get input in binary form

#             # Separate the inputs
#             a = int(input_bin[0])
#             b = int(input_bin[1])

#             # Perform the half subtractor operation (difference and borrow)
#             diff_bit = a ^ b          # Difference bit
#             borrow_bit = (not a) & b  # Borrow bit

#             # Combine borrow and difference into a single binary number
#             output_bin = f'{borrow_bit}{diff_bit}'

#             # Mask bits (equal to number of output bits)
#             mask = "11"

#             # Write input, output, and mask to the trace file
#             optr.write(f'{input_bin} {output_bin} {mask}\n')

# # Run the main function
# if __name__ == "__main__":
#     generate_tracefile()

